streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java

```diff
@@ -205,7 +209,9 @@ void restoreAllInternal(final Collection<KeyValue<byte[], byte[]>> records) {
             for (final Map.Entry<KeyValueSegment, WriteBatch> entry : writeBatchMap.entrySet()) {
                 final KeyValueSegment segment = entry.getKey();
                 final WriteBatch batch = entry.getValue();
-                segment.write(batch);
+                if (segment.open) {
+                    segment.write(batch);
+                }
```

Comment:
Can you elaborate? Seems to be orthogonal to the timestamp fix.

Reply:
The segmented store expires segments based on stream time.

Previously, the stream time it used was coming from the ProcessorContext, which doesn't get updated while we're processing this batch write, so the whole method execution happened in one "instant". That instant would be at least the latest time in the batch. So when it loaded all the segments in building `batchWriteMap`, it was already able to tell which records were for expired segments and drop them.

Now, though, it tracks stream time based on the sequence of received writes, so at the beginning of this method call, the local stream time is maybe at the beginning of the batch. This means that in the beginning of this method, it creates/gets segments for everything in the batch (not updating the stream time at all), so all the segments are still live until we enter the loop over `writeBatchMap`.

When we loop over the batch map, we're actually processing the writes, which advances the local view of stream time, possibly expiring some segments. Since stream-time is actually advancing during the loop, and segments are being expired as the loop executes, we can no longer expect all our segments to be open here.

Reply:
Thanks for the details. Should we instead do a "clean up" within `getWriteBatches()` and only return open segments? Seems to be a better separation of concerns.

Reply:
Hah. Ok, that's what I get for commenting from memory. It turns out that we actually do advance time within `getWriteBatches`. The problem is actually that in that method, we're advancing time while we're also creating the segments and populating the result list. So, by the end of the iteration, some of the segments at the beginning of the result list may already be expired.

We can fix it by either pre-iterating over all the records to advance time to the largest timestamp in the batch or post-iterating over the results to remove the expired segments.

I opted for the pre-iteration so that we can potentially avoid creating and opening a segment that would just need to be closed and cleaned up later in the same batch.

The code is cleaned up now. Thanks for the catch!